fred2/missioncutscenesdlg.cpp

```diff
@@ -0,0 +1,466 @@
+{
+	m_cutscenes_tree.right_clicked(MODE_CUTSCENES); // We don't need to perform actions here, so use the same method as Goals
+	*pResult = 0;
+}
+
+// cutscene tree item label editing is requested.  Determine if it should be allowed.
+void CMissionCutscenesDlg::OnBeginlabeleditCutscenesTree(NMHDR* pNMHDR, LRESULT* pResult)
+{
+	TV_DISPINFO* pTVDispInfo = (TV_DISPINFO*)pNMHDR;
+
+	if (m_cutscenes_tree.edit_label(pTVDispInfo->item.hItem) == 1) {
+		*pResult = 0;
+		modified = 1;
+	} else {
+		*pResult = 1;
+	}
+}
+
+// Once we finish editing, we need to clean up, which we do here.
+void CMissionCutscenesDlg::OnEndlabeleditCutscenesTree(NMHDR* pNMHDR, LRESULT* pResult)
+{
+	TV_DISPINFO* pTVDispInfo = (TV_DISPINFO*)pNMHDR;
+
+	*pResult = m_cutscenes_tree.end_label_edit(pTVDispInfo->item);
+}
+
+void CMissionCutscenesDlg::OnOK()
+{
+	CWnd *w;
+
+	w = GetFocus();
+	if (w) {
+		HWND h = w->m_hWnd;
+		GetDlgItem(IDC_CUTSCENES_TREE)->SetFocus();
+		::SetFocus(h);
+	}
+}
+
+int CMissionCutscenesDlg::query_modified()
+{
+	if (modified)
+		return 1;
+
+	if (The_mission.cutscenes.size() != m_cutscenes.size())
+		return 1;
+
+	for (int i = 0; i < (int)The_mission.cutscenes.size(); i++) {
+		if (!stricmp(The_mission.cutscenes[i].filename, m_cutscenes[i].filename))
+			return 1;
+		if (The_mission.cutscenes[i].type != m_cutscenes[i].type)
+			return 1;
+	}
+
+	return 0;
+}
+
+void CMissionCutscenesDlg::OnOk()
+{
+	SCP_vector<std::pair<SCP_string, SCP_string>> names;
+
+	UpdateData(TRUE);
+	if (query_modified())
+		set_modified();
+
+	for (auto& cutscene : The_mission.cutscenes) {
+		free_sexp2(cutscene.formula);
+	}
+
+	// copy all dialog cutscenes to the mission
+	The_mission.cutscenes.clear();
+	for (const auto& dialog_cutscene : m_cutscenes) {
+		The_mission.cutscenes.push_back(dialog_cutscene);
+		The_mission.cutscenes.back().formula = m_cutscenes_tree.save_tree(dialog_cutscene.formula);
+	}
+
+	theApp.record_window_data(&Mission_cutscenes_wnd_data, this);
+	CDialog::OnOK();
+}
+
+void CMissionCutscenesDlg::OnButtonNewCutscene()
+{
+	m_cutscenes.emplace_back();
+	m_sig.push_back(-1);
+
+	m_cutscenes.back().type = m_display_cutscene_types;
+	strcpy_s(m_cutscenes.back().filename, "cutscene filename");
+	HTREEITEM h = m_cutscenes_tree.insert(m_cutscenes.back().filename);
+
+	m_cutscenes_tree.item_index = -1;
+	m_cutscenes_tree.add_operator("true", h);
+	int index = m_cutscenes.back().formula = m_cutscenes_tree.item_index;
+	m_cutscenes_tree.SetItemData(h, index);
+
+	m_cutscenes_tree.SelectItem(h);
+}
+
+int CMissionCutscenesDlg::handler(int code, int node)
+{
+	switch (code) {
+		case ROOT_DELETED:
+			int i;
+			for (i = 0; i < (int)m_cutscenes.size(); i++){
+				if (m_cutscenes[i].formula == node) {
+					break;
+				}
+			}
+
+			Assert(i < (int)m_cutscenes.size());
+			m_cutscenes.erase(m_cutscenes.begin() + i);
+			m_sig.erase(m_sig.begin() + i);
+
+			return node;
+
+		default:
+			Int3();
```

Comment:
Casts, Assertion instead of Assert, and Int3()

You can use UNREACHABLE() instead of Int3(), which allows a message to be displayed to the user, like Assertion does.